packages/labs/src/inputList.tsx

```diff
@@ -0,0 +1,293 @@
+    }
+
+    public componentWillReceiveProps(nextProps: IInputListProps<T>) {
+        if (nextProps.items !== this.props.items
+            || nextProps.itemListPredicate !== this.props.itemListPredicate
+            || nextProps.itemPredicate !== this.props.itemPredicate
+            || nextProps.query !== this.props.query
+        ) {
+            this.shouldCheckActiveItemInViewport = true;
+            this.setState({ filteredItems: getFilteredItems(nextProps) });
+        }
+    }
+
+    public componentDidUpdate() {
+        if (this.shouldCheckActiveItemInViewport) {
+            // update scroll position immediately before repaint so DOM is accurate
+            // (latest filteredItems) and to avoid flicker.
+            requestAnimationFrame(() => this.scrollActiveItemIntoView());
+            // reset the flag
+            this.shouldCheckActiveItemInViewport = false;
+        }
+        // reset active item (in the same step) if it's no longer valid
+        if (this.getActiveIndex() < 0) {
+            Utils.safeInvoke(this.props.onActiveItemChange, this.state.filteredItems[0]);
+        }
+    }
+
+    public scrollActiveItemIntoView() {
+        const activeElement = this.getActiveElement();
+        if (this.itemsParentRef != null && activeElement != null) {
+            const { offsetTop: activeTop, offsetHeight: activeHeight } = activeElement;
+            const {
+                offsetTop: parentOffsetTop,
+                scrollTop: parentScrollTop,
+                clientHeight: parentHeight,
+            } = this.itemsParentRef;
+            // compute padding on parent element to ensure we always leave space
+            const { paddingTop, paddingBottom } = this.getItemsParentPadding();
+
+            // compute the two edges of the active item for comparison, including parent padding
+            const activeBottomEdge = activeTop + activeHeight + paddingBottom - parentOffsetTop;
+            const activeTopEdge = activeTop - paddingTop - parentOffsetTop;
+
+            if (activeBottomEdge >= parentScrollTop + parentHeight) {
+                // offscreen bottom: align bottom of item with bottom of viewport
+                this.itemsParentRef.scrollTop = activeBottomEdge + activeHeight - parentHeight;
+            } else if (activeTopEdge <= parentScrollTop) {
+                // offscreen top: align top of item with top of viewport
+                this.itemsParentRef.scrollTop = activeTopEdge - activeHeight;
+            }
+        }
+    }
+
+    private getActiveElement() {
+        if (this.itemsParentRef != null) {
+            return this.itemsParentRef.children.item(this.getActiveIndex()) as HTMLElement;
+        }
+        return undefined;
+    }
+
+    private getActiveIndex() {
+        // NOTE: this operation is O(n) so it should be avoided in render(). safe for events though.
+        return this.state.filteredItems.indexOf(this.props.activeItem);
```

Comment:
nit: this function gets called multiple times every time the user presses up/down and may lead to a laggy experience if the filteredItems size is large and the user holds up/down. May be able to make this faster with something like an ordered map but probably not worth the trouble if there are plans to virtualize this. Actually, if the number of items is large, the rendering perf may dwarf whatever perf issue this may result in...